[JENKINS-22474] API Token does not require CSRF token

Wadeck opened the pull request:
- in order to ease the use of the api, we are not requiring the request to have a crumb
- in terms of security it's not a problem normally since the CSRF attacks use the cookies and in case of API Token, it's session-less / cookie-less

See [JENKINS-22474](https://issues.jenkins-ci.org/browse/JENKINS-22474).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Crumb is no more required for API calls using API Token

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees @jglick 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

Wadeck committed:
[JENKINS-22474] API Token does not require CSRF token
- in order to ease the use of the api, we are not requiring the request to have a crumb
- in terms of security it's not a problem normally since the CSRF attacks use the cookies and in case of API Token, it's session-less / cookie-less

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

recena commented on core/src/main/java/jenkins/security/ApiCrumbExclusion.java:
```diff
@@ -0,0 +1,54 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 CloudBees, Inc.
```
`2017`

Wadeck committed:
- adjust the license header

daniel-beck commented:
Needs testing with https://wiki.jenkins.io/display/JENKINS/Reverse+Proxy+Auth+Plugin (and basic auth) to ensure CSRF protection on the UI is still effective with that.

Wadeck committed:
- add test for basic authentication
- add test for login process

Wadeck committed:
- add test for form submission using ui (htmlunit), not just login form

Wadeck commented:
@daniel-beck test added for basic auth / login form / ui form submission
I will test using the plugin manually (or there is a way to include a test for that ?)

Wadeck commented:
@daniel-beck results of manual testing after installing the plugin:
- with Postman for the REST: crumb is mandatory in any case (as we do not pass through the `BasicHeaderApiTokenAuthenticator`), either with API Token or HTTP header
- with ModHeader chrome extension to add the header requested by the plugin, CSRF check is done like before (i.e. crumb mandatory)

jglick commented on core/src/main/java/jenkins/security/ApiCrumbExclusion.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    public boolean process(HttpServletRequest request, HttpServletResponse response, FilterChain chain) throws IOException, ServletException {
+        Object apiTokenValid = request.getAttribute(BasicHeaderApiTokenAuthenticator.class.getName());
+        if (apiTokenValid != null && apiTokenValid == Boolean.TRUE) {
```
i.e.,

```java
if (Boolean.TRUE.equals(request.getAttribute(BasicHeaderApiTokenAuthenticator.class.getName())) {
```

Wadeck replied:
modified as requested

jglick reviewed (approved):
Main code change seems fine. You should take the opportunity to clean up test cases which currently pass a crumb just in order to make a REST request (typically using password-based login, which is slow since it loads a form in HtmlUnit) with a more realistic call. For example:

```diff
diff --git a/test/src/test/java/hudson/model/ProjectTest.java b/test/src/test/java/hudson/model/ProjectTest.java
index a2d5bf3f67..e7f8b4a8a1 100644
--- a/test/src/test/java/hudson/model/ProjectTest.java
+++ b/test/src/test/java/hudson/model/ProjectTest.java
@@ -74,6 +74,7 @@ import java.util.List;
 import java.util.ArrayList;
 import java.io.IOException;
 import java.nio.charset.Charset;
+import java.util.Base64;
 import java.util.Collection;
 import org.junit.Rule;
 import org.junit.Test;
@@ -87,6 +88,7 @@ import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.logging.Logger;
 import jenkins.model.BlockedBecauseOfBuildInProgress;
+import jenkins.security.ApiTokenProperty;
 
 import org.junit.Ignore;
 import org.jvnet.hudson.test.Issue;
@@ -583,7 +585,6 @@ public class ProjectTest {
         FreeStyleProject project = j.createFreeStyleProject("project");
         GlobalMatrixAuthorizationStrategy auth = new GlobalMatrixAuthorizationStrategy();   
         j.jenkins.setAuthorizationStrategy(auth);
-        j.jenkins.setCrumbIssuer(null);
         HudsonPrivateSecurityRealm realm = new HudsonPrivateSecurityRealm(false);
         j.jenkins.setSecurityRealm(realm); 
         User user = realm.createAccount("John Smith", "password");
@@ -605,7 +606,8 @@ public class ProjectTest {
         String cmd = "echo hello > change.log";
         project.getBuildersList().add(Functions.isWindows()? new BatchFile(cmd) : new Shell(cmd));
         j.buildAndAssertSuccess(project);
-        JenkinsRule.WebClient wc = j.createWebClient().login(user.getId(), "password");
+        JenkinsRule.WebClient wc = j.createWebClient();
+        wc.addRequestHeader("Authorization", "Basic " + Base64.getEncoder().encodeToString((user.getId() + ":" + user.getProperty(ApiTokenProperty.class).getApiToken()).getBytes()));
         WebRequest request = new WebRequest(new URL(wc.getContextPath() + project.getUrl() + "doWipeOutWorkspace"), HttpMethod.POST);
         HtmlPage p = wc.getPage(request);
         Thread.sleep(500);
```

except of course we should enhance `jenkins-test-harness` to make this more convenient, so that `WebClient` would have methods like

```java
public WebClient usingToken(User user) {
    addRequestHeader("Authorization", "Basic " + Base64.getEncoder().encodeToString((user.getId() + ":" + user.getProperty(ApiTokenProperty.class).getApiToken()).getBytes()));
    return this;
}
public WebClient usingToken(String userID) {
    return usingToken(User.getById(userID));
}
```

jglick commented:
In general, review usages of `createCrumbedUrl` and `addCrumb` and you will find plenty of tests which could be expressed both more realistically and more simply with such a `usingToken` utility in this PR.

Wadeck commented:
Related PRs on reverse-proxy-auth-plugin: [#31 for doc](https://github.com/jenkinsci/reverse-proxy-auth-plugin/pull/31), [#32 for using User.getById instead of User.get](https://github.com/jenkinsci/reverse-proxy-auth-plugin/pull/32)

Wadeck committed:
- modification requested by Jesse

Wadeck commented:
@jglick Your request concerning the usingToken instead of login seems to be very effective. Some tests go from 30sec to only 10.

Wadeck committed:
- pom.xml update to use the last version of jenkins-test-harness (with the token helper methods)
- beginning of the simplification of tests

Wadeck committed:
- using the try-with-resource approach to ease readability

Wadeck committed:
- using closure method now

Wadeck committed:
- add missing login transformation

Wadeck committed:
- add unit test

Wadeck committed:
- use withToken
- remove useless crumb for GET method
- add fail (otherwise the assert in catch is not as useful as it could be)

Wadeck committed:
another bunch of test cases

Wadeck committed:
- for HudsonTestCase, some additional modifications are required: changing the view / different type of management for the variable inside the views

Wadeck committed:
- small other tests

Wadeck committed:
- last batch for the login method

Wadeck committed:
- crumb is not more required since we are using API Token

Wadeck committed:
- converting auth to ApiToken to avoid crumb method

Wadeck committed:
- converting auth to ApiToken to avoid crumb method (second)

Wadeck commented:
⚠️ Version of JTH used is a local snapshot (reason of the test failing), as soon as a new release is done, the PR will be modified.

Wadeck committed:
- remove usage of closure aware methods

oleg-nenashev commented:
On hold till JTH gets released and updated

Wadeck committed:
- update the pom using the snapshot as adviced by Jesse
- modifications on other class to adapt to the last modifications in JTH

Wadeck commented:
JTH updated and snapshot deployed, related code updated.

jglick commented on test/pom.xml:
```diff
@@ -53,7 +53,9 @@ THE SOFTWARE.
-      <version>2.31</version>
+      <!--<version>2.31</version>-->
+      <!--<version>2.32-SNAPSHOT</version>-->
+      <version>2.32-20171113.154040-2</version>
```
2.32 is already released; please merge with `master` and update to it.

jglick commented on core/src/main/java/jenkins/security/ApiCrumbExclusion.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * JENKINS-22474: Makes API Token calls bypass CSRF protection to ease usage
+ */
+@Symbol("apiCrumbExclusion")
```
Symbols should not contain the name of the extension point, only that part which distinguishes them from other implementations of the same extension point. So for example `apiToken` would be appropriate here.

jglick commented on test/src/test/java/hudson/bugs/JnlpAccessWithSecuredHudsonTest.java:
```diff
@@ -86,7 +87,7 @@ protected Slave createNewJnlpSlave(String name) throws Exception {
         r.jenkins.setNodes(Collections.singletonList(createNewJnlpSlave("test")));
         JenkinsRule.WebClient wc = r.createWebClient();
-        HtmlPage p = wc.login("alice").goTo("computer/test/");
+        HtmlPage p = wc.withBasicApiToken(User.getById("alice", true)).goTo("computer/test/");
```
Why does simply `withBasicApiToken("alice")` not work here? Does that user not exist until created here?

Wadeck replied:
=> `java.lang.AssertionError: The userId must correspond to an already created User` if I just put Alice's login. The user must be created before using the methd and is not yet created.

jglick commented on test/src/test/java/hudson/model/AbstractProject2Test.java:
```diff
@@ -56,16 +56,20 @@ public void upstreamDownstreamExportApi() throws Exception {
             grant(Item.READ).onItems(ds).to("charlie"));
-        String api = r.createWebClient().login("alice").goTo(us.getUrl() + "api/json?pretty", null).getWebResponse().getContentAsString();
+
+        String api = r.createWebClient().withBasicCredentials("alice", "alice").goTo(us.getUrl() + "api/json?pretty", null).getWebResponse().getContentAsString();
```
I thought you could use just `withBasicCredentials("alice")` if the password was the same, as it is for `DummySecurityRealm`?

jglick commented on test/src/test/java/hudson/model/ItemsTest.java:
```diff
@@ -61,6 +62,10 @@
     @Rule public TemporaryFolder tmpRule = new TemporaryFolder();
 
+    @Before public void prepareAttackerUser(){
+        User.getById("attacker", true);
```
Move to `overwriteTargetSetUp`.

jglick commented on test/src/test/java/hudson/model/ItemsTest.java:
```diff
@@ -185,7 +190,7 @@ private void cannotOverwrite(String target) throws Exception {
                 wc.getOptions().setRedirectEnabled(false);
                 wc.getOptions().setThrowExceptionOnFailingStatusCode(false); // redirect perversely counts as a failure
-                WebResponse webResponse = wc.getPage(new WebRequest(createCrumbedUrl(r, wc, "createItem?name=" + target + "&mode=hudson.model.FreeStyleProject"), HttpMethod.POST)).getWebResponse();
+                WebResponse webResponse = wc.getPage(new WebRequest(new URL(wc.getContextPath() + "createItem?name=" + target + "&mode=hudson.model.FreeStyleProject"), HttpMethod.POST)).getWebResponse();
```
BTW we could probably introduce `WebClient.postTo(String)` (or something) as a convenience.

jglick commented on test/src/test/java/hudson/model/ProjectTest.java:
```diff
@@ -568,7 +571,12 @@ public void testDoDoDelete() throws Exception{
-        List<HtmlForm> forms = j.createWebClient().login(user.getId(), "password").goTo(project.getUrl() + "delete").getForms();
+
+        JenkinsRule.WebClient wc = j.createWebClient();
+        wc.withBasicCredentials(user.getId(), "password");
```
BTW the usage of `HudsonPrivateSecurityRealm` here is probably unnecessary; `DummySecurityRealm` would be simpler.

jglick commented on test/src/test/java/hudson/model/ProjectTest.java:
```diff
@@ -568,7 +571,12 @@ public void testDoDoDelete() throws Exception{
+
+        JenkinsRule.WebClient wc = j.createWebClient();
+        wc.withBasicCredentials(user.getId(), "password");
+        HtmlPage p = wc.goTo(project.getUrl() + "delete");
```
Note that use of BASIC credentials here is less realistic than the original version, since this is testing HTML UI that would in practice always be preceded by a cookie-based login. For purposes of this test it probably does not matter and the speedup may be worth the slight loss of fidelity.

jglick commented on test/src/test/java/hudson/security/HudsonPrivateSecurityRealmTest.java:
```diff
@@ -67,6 +67,7 @@ public void dataCompatibilityWith1_282() throws Exception {
             // verify the sanity that the password is really used
             // this should fail
             j.createWebClient().login("bob", "bob");
+            fail();
```
Independent I guess, but good catch.

Wadeck replied:
Yeah I saw some tests like that, if the code did not fail, the test is still passing...

jglick commented on test/src/test/java/hudson/util/RobustReflectionConverterTest.java:
```diff
@@ -26,17 +26,11 @@
-import hudson.model.FreeStyleProject;
-import hudson.model.Job;
-import hudson.model.Saveable;
+import hudson.model.*;
```
Please avoid star imports.

jglick commented on test/src/test/java/jenkins/model/JenkinsTest.java:
```diff
@@ -338,18 +341,22 @@ public void testDoEval() throws Exception {
         }
-        wc.login("charlie");
+
+        wc.withBasicApiToken(User.getById("charlie", true));
```
Maybe it would be better for `withBasicApiToken(String)` to just call `getById(name, true)`? If you really wanted to assert that the user record already existed (which for a typical test using `DummySecurityRealm` would be irrelevant), you could do so explicitly. It would simplify a bunch of code in this PR to make such a change.

Wadeck replied:
Initially my idea was that could lead to potential misunderstanding. We could change that behavior in a second PR in order to have this one merged.

jglick replied:
Agreed.

jglick commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -111,28 +94,53 @@ public void testAuthHeaderCaseInSensitive() throws Exception {
+        if (userAndPass==null) {
+            return null;
+        }
+        return prefix+" "+Scrambler.scramble(userAndPass);
```
Use `Base64`. `Scrambler` should be considered deprecated.

Wadeck replied:
As you showed me the existence of `java.util.Base64` in Java 8, I think you are talking about that one.

jglick commented on test/src/test/java/jenkins/security/ApiCrumbExclusionTest.java:
```diff
@@ -0,0 +1,146 @@
+        req.setHttpMethod(HttpMethod.POST);
+        req.setEncodingType(null);
+        Page p = wc.getPage(req);
+        assertEquals(expected, p.getWebResponse().getContentAsString().trim());
```
BTW `trim` should be unnecessary now that we are using `text` not `plainText`.

jglick commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -111,28 +94,53 @@ public void testAuthHeaderCaseInSensitive() throws Exception {
+        if (authCode!=null)
+            req.setAdditionalHeader("Authorization", authCode);
+        Page p = wc.getPage(req);
+        assertEquals(expectedLogin, p.getWebResponse().getContentAsString().trim());
```
ditto, could drop `trim`

jglick commented on test/src/test/java/lib/form/PasswordTest.java:
```diff
@@ -62,25 +57,43 @@
 
-    public void test1() throws Exception {
-        secret = Secret.fromString("secret");
-        HtmlPage p = createWebClient().goTo("self/test1");
```
Oh thank you, these `self` references from `HudsonTestCase`s were a pesky impediment to switching to `JenkinsRule`.

jglick commented on test/src/test/resources/lib/form/PasswordTest/SecretNotPlainText/index.jelly:
```diff
@@ -35,6 +35,13 @@ THE SOFTWARE.
+      </f:entry>
+      <f:entry field="titi">
+        <f:textbox name="parameter.defaultValue2" value="${instance.titi}" />
+      </f:entry>
```
These new fields seem unused?

jglick reviewed (approved):
Some suggestions about tests, but patch seems fine to me. Really the main question is: have you demonstrated to your own satisfaction that no CSRF attack is possible which would obtain the API token and then pass it in a basic auth header? Assuming so, we should be good to go.

jglick commented:
(other than the `jenkins-test-harness` snapshot version)

Wadeck committed:
Merge remote-tracking branch 'upstream/master' into JENKINS-22474_CRUMB_POST_WITH_API_TOKEN

Wadeck committed:
- modifications requested during code review

oleg-nenashev commented on test/pom.xml:
```diff
@@ -53,7 +53,7 @@ THE SOFTWARE.
       <groupId>${project.groupId}</groupId>
       <artifactId>jenkins-test-harness</artifactId>
-      <version>2.31</version>
+      <version>2.32</version>
```
would be better to use a dependencyManagement

jglick replied:
Why? There is exactly one dependency on this artifact from this repository.

oleg-nenashev replied:
Do not recall why I said that

oleg-nenashev reviewed (approved):
The implementation looks good to me.
I didn't read all the test changes, IIUC there are some unrelated changes.It is fine to keep it as is, because I doubt this change would be an LTS candidate.

oleg-nenashev commented:
Merge conflict needs to be resolved.
I also think a feedback from @daniel-beck is needed

Wadeck committed:
Merge remote-tracking branch 'upstream/master' into JENKINS-22474_CRUMB_POST_WITH_API_TOKEN

# Conflicts:
#	test/src/test/java/hudson/model/AbstractProject2Test.java

Wadeck committed:
- also put back my changes to the conflicted file

oleg-nenashev commented:
@Wadeck could you please resolve the merge conflict?

Wadeck committed:
Merge branch 'master' into JENKINS-22474_CRUMB_POST_WITH_API_TOKEN

Wadeck committed:
- correction of the merge :)

jglick commented:
Unfortunately the base merge made the incremental diff in GH useless, so will need to look at the full diff…
